### PR TITLE
[quant][core] Add quantize/dequantize ops for decomposed quantized Tensor representation

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87093

Summary:
Added q/dq implementation for out of core (decomposed) quantized Tensor representation, meaning that
instead of storing quantization parameters (e.g. scale/zero_point) in a separate quantized Tensor object, we will store
quantization parameters in the argument of operators.
```
quantize(float32_tensor, scale, zero_point, dtype) -> int8_tensor
dequantize(int8_tensor, scale, zero_point, dtype) -> float32_tensor
```

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_decomposed_quantize
python test/test_quantization.py TestQuantizedTensor.test_decomposed_dequantize

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: aten/src/ATen/native/quantized/decomposed/DeQuantize.cpp
Comment (dzdang) [inline aten/src/ATen/native/quantized/decomposed/DeQuantize.cpp:10]: is the `.to` needed here when `* scale` implicitly converts the LHS to a double?

**Review by dzdang — APPROVED** [review]
lgtm

File: aten/src/ATen/native/quantized/decomposed/DeQuantize.cpp
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/decomposed/DeQuantize.cpp:10]: looks like so, the test failed when we don't have this..

File: aten/src/ATen/native/quantized/decomposed/Quantize.cpp
Comment (vkuzo) [inline aten/src/ATen/native/quantized/decomposed/Quantize.cpp:7]: would it be more forward proof to have `qmin` and `qmax` in the signature?

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/decomposed/Quantize.cpp:7]: yeah I was thinking about this as well, should I add it in this PR?

Comment (vkuzo) [reply aten/src/ATen/native/quantized/decomposed/Quantize.cpp:7]: I think that would be good, it might be annoying to change the signature later as I think there are tests that check for BC compatibility of signatures of functions in native_functions.yaml

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/decomposed/Quantize.cpp:7]: OK, sg, will update the PRs then

**Comment by ezyang** [comment]
Why not do this in Python?

**Comment by jerryzh168** [comment]
> Why not do this in Python?

I think adding this to python means we will just see more complicated patterns for quantized operators after exir.capture, e.g.
`(x - zero_point) + scale -> linear -> x / scale + zero_point ` instead of `dq -> linear -> q`, looks like this is more aligned with long term plan anyways, maybe we should do that now actually

**Comment by ezyang** [comment]
you can also register it as CompositeExplicitAutograd from Python which will prevent it from showing up in trace. (need to write derivative formulas etc, if needed)

**Comment by jerryzh168** [comment]
> 

I see, is this how we do that? https://colab.research.google.com/drive/1RRhSfk7So3Cn02itzLWE9K4Fam-8U011#scrollTo=RulQ0zou-SGH

**Comment by ezyang** [comment]
this works too

**Comment by jerryzh168** [comment]
> this works too

updated, please take a look again

File: torch/ao/quantization/fx/_decomposed.py
Comment (ezyang) [inline torch/ao/quantization/fx/_decomposed.py:25]: You can use fstrings here

File: torch/ao/quantization/fx/_decomposed.py
Comment (ezyang) [inline torch/ao/quantization/fx/_decomposed.py:39]: this is not an f-string lol

File: torch/ao/quantization/fx/_decomposed.py
Comment (ezyang) [inline torch/ao/quantization/fx/_decomposed.py:37]: Your implementations here are not actually CPU specific, so I'd recommend registering them as "CompositeExplicitAutograd" instead

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:39]: oh sorry.. will change :)

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:37]: do I need to implement backward in that case?

Comment (ezyang) [reply torch/ao/quantization/fx/_decomposed.py]: yes, if you want to differentiate through it.

Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py]: OK, I don't want to differentiate it, so will skip that for now

File: torch/ao/quantization/fx/_decomposed.py
Comment (vkuzo) [inline torch/ao/quantization/fx/_decomposed.py:38]: thoughts on why `quant_min` and `quant_max` are in this function?  They are necessary for quantizing, but not sure why they would be needed for dequantizing.

File: torch/ao/quantization/fx/_decomposed.py
Comment (vkuzo) [inline torch/ao/quantization/fx/_decomposed.py:41]: maybe we can subtract zero_point before casting to float?

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:38]: I think it might be useful as an metadata for the quantized Tensor input, that could be useful for pattern matching in the future, but I don't have a strong opinion on this one.

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:41]: sure, that makes sense

File: torch/ao/quantization/fx/_decomposed.py
Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: sorry, what do you mean? not sure I understand yet

Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:38]: sure, let me give two examples:

(1). use quant_min/quant_max to show the range restriction of input
pattern for xnnpack linear (simplified):
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # range restriction for weight
       w = dequantize(x, scale, zero_point, -127, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x

(2). using int32 to simulate int16
pattern for a int16 op
def forward(self, x):
    # we use quant_min/quant_max in dequantize to show that we have an int16 input tensor
    x = dequantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    return x

File: torch/ao/quantization/fx/_decomposed.py
Comment (ezyang) [inline torch/ao/quantization/fx/_decomposed.py:4]: Are you sure you want to name this "quantized_decomposed"? Doesn't seem very descriptive

**Comment by ezyang** [comment]
I can't really comment on the algos, but the use of python op registration looks legit.

File: torch/ao/quantization/fx/_decomposed.py
Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:41]: hmm, sorry, just to clarify, we plan to fix this before landing, right?

File: torch/ao/quantization/fx/_decomposed.py
Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: I understand why quant_min/quant_max are needed for quantize, still don't understand why they are there for dequantize.  Is there some need to use them inside of the dequantize function in the future?  Not sure why their presence or absence would affect pattern matching either.

Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: is it just carrying the metadata of "this was quantized with quant_min=a and quant_max=b" around?

Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: I think not having quant_min/quant_max in dequantize makes more sense, but don't want to block on that.  The rest of the PR LGTM.

Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: > in the new design, quant_min and quant_max might be useful to decide which op the pattern corresponds to, as is shown in the example.

>  I think an alternative can be we traverse back one more step to quantize op to find the quant_min/quant_max metadata for the input of dequantize op,

I understand what you're saying, still missing on why this is needed though.  Could you provide an example where quant_min/quant_max on the quantize is not enough?  I'm not sure that the examples above demonstrate that.

Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:38]: Are you trying to say that there will be future logic which would do something like

```
    # we know the int tensor satisfies the constraint of -2**15 <= int_val <= 2**15 - 1, so we can lower to op1
    x = dequantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)

    # we know the int tensor satisfies the constraint of -2**15 <= int_val <= 2**15 - 1, 
    # so we cannot lower to op1 without more information
    x = dequantize(x, scale, zero_point, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
```
?

Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:38]: Sure, let's say we have the following ops in the same backend:

* op1: xnnpack linear
** activation dtype = int8, range=-128, 127
** weight dtype = int8, range=**-128**, 127
* op2 another linear implementation
** activation dtype = int8, range=-127, 127
** weight dtype = int8, range=**-127**, 127

we'll need the following two patterns to match them:
## op1 pattern:
```
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # range restriction for weight
       w = dequantize(x, scale, zero_point, -127, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x
```

## op2 pattern:
```
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # full range restriction for weight
       w = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x
```

Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:38]: yeah, I'm also open to removing it, if there are no such use cases, sg, I'll document this and land for now, we should be able to revisit the decision later if needed, since the reference quantized model are not serialized so we should be able to change that

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:47]: I tried changing the order but it turns out that it failed the test (comparing with existing dequant implementation), I'm planning to land for now to unblock and we can investigate better, wdyt?

Comment (vkuzo) [reply torch/ao/quantization/fx/_decomposed.py:47]: sure

File: torch/ao/quantization/fx/_decomposed.py
Comment (larryliu0820) [inline torch/ao/quantization/fx/_decomposed.py:9]: So we need to register exactly the same schema somewhere into edge runtime. It seems a bit hard to maintain for current infra. If we change schema here we need to change schema in edge runtime otherwise it won't work.

File: torch/ao/quantization/fx/_decomposed.py
Comment (z-a-f) [inline torch/ao/quantization/fx/_decomposed.py:48]: nit: I think this fails because of the under-/overflow? For example, if `dtype(input) == uint8`, `input == 0`, and `zero_point > 0`, you will have wrapped around values:

```
>>> import torch
>>> a = torch.tensor([1,2,3], dtype=torch.uint8)
>>> a - 3
tensor([254, 255,   0], dtype=torch.uint8)
```

File: torch/ao/quantization/fx/_decomposed.py
Comment (z-a-f) [reply torch/ao/quantization/fx/_decomposed.py:6]: This name is giving me jitters :)

**Review by z-a-f — APPROVED** [review]

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:9]: so should we define these in native_functions.yaml?

File: torch/ao/quantization/fx/_decomposed.py
Comment (jerryzh168) [reply torch/ao/quantization/fx/_decomposed.py:48]: yeah could be, but I remember seeing fbgemm is doing the same: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtils.h#L139-L141, or maybe the subtraction in `(src - qparams.zero_point)` is an int32 since qparams.zero_point is int32: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtilsAvx2.h#L21

**Pull request closed without merging**
